### ISSUE TITLE
Closure does not optimize redundant 'var's, but even introduces new 'var' keywords in inner loop

### ISSUE THREAD

**Issue by juj** [google/closure-compiler#3789]
Optimizing Emscripten WebGL runtime library in https://github.com/emscripten-core/emscripten/pull/13732, I have a for loop:

```js
for (i = 0; i < GLctx.getProgramParameter(program, 0x8B86/*GL_ACTIVE_UNIFORMS*/); ++i) {
  var u = GLctx.getActiveUniform(program, i);
  var nm = u.name;
  var sz = u.size;
  var lb = getLeftBracePos(nm);
  var arrayName = lb > 0 ? nm.slice(0, lb) : nm;

  // Assign a new location.
  var id = program.uniformIdCounter;
  program.uniformIdCounter += sz;

  // Eagerly get the location of the uniformArray[0] base element.
  // The remaining indices >0 will be left for lazy evaluation to
  // improve performance. Those may never be needed to fetch, if the
  // application fills arrays always in full starting from the first
  // element of the array.
  uniformSizeAndIdsByName[arrayName] = [sz, id];

  // Store placeholder integers in place that highlight that these
  // >0 index locations are array indices pending population.
  for(j = 1; j < sz; ++j) {
    uniformLocsById[++id] = j;
    program.uniformArrayNamesById[id] = arrayName;
  }
}
```
When Closure goes through this, it emits
```js
for (a.P = d = {}, a.T = {}, h = 0; h < P.getProgramParameter(a, 35718); ++h) {
    var n = P.getActiveUniform(a, h);
    var r = n.name;
    n = n.size;
    var w = c(r);
    w = 0 < w ? r.slice(0, w) : r;
    var F = a.R;
    a.R += n;
    e[w] = [n, F];
    for (r = 1; r < n; ++r) d[++F] = r, a.T[F] = w
}
```
Notice the four copies of `var` keyword. One can further optimize by hand to
```js
for (a.P = d = {}, a.T = {}, h = 0; h < P.getProgramParameter(a, 35718); ++h) {
    var n = P.getActiveUniform(a, h),
    r = n.name,
    n = n.size,
    w = c(r),
    F = a.R;
    w = 0 < w ? r.slice(0, w) : r;
    a.R += n;
    e[w] = [n, F];
    for (r = 1; r < n; ++r) d[++F] = r, a.T[F] = w
}
```
saving 3x`var `==12 bytes.

Attempting this optimization in source code:
```js
for (i = 0; i < GLctx.getProgramParameter(program, 0x8B86/*GL_ACTIVE_UNIFORMS*/); ++i) {
  var u = GLctx.getActiveUniform(program, i),
    nm = u.name,
    sz = u.size,
    lb = getLeftBracePos(nm),
    arrayName = lb > 0 ? nm.slice(0, lb) : nm,
    id = program.uniformIdCounter;

  program.uniformIdCounter += sz;

  // Eagerly get the location of the uniformArray[0] base element.
  // The remaining indices >0 will be left for lazy evaluation to
  // improve performance. Those may never be needed to fetch, if the
  // application fills arrays always in full starting from the first
  // element of the array.
  uniformSizeAndIdsByName[arrayName] = [sz, id];

  // Store placeholder integers in place that highlight that these
  // >0 index locations are array indices pending population.
  for(j = 1; j < sz; ++j) {
    uniformLocsById[++id] = j;
    program.uniformArrayNamesById[id] = arrayName;
  }
}
```
results in Closure undoing the optimization and reintroducing the `var`s:
```js
for (a.P = d = {}, a.T = {}, h = 0; h < P.getProgramParameter(a, 35718); ++h) {
    var n = P.getActiveUniform(a, h);
    var r = n.name;
    n = n.size;
    var w = c(r);
    w = 0 < w ? r.slice(0, w) : r;
    var F = a.R;
    a.R += n;
    e[w] = [n, F];
    for (r = 1; r < n; ++r) d[++F] = r, a.T[F] = w
}
```

Moving the declaration of the `var`s outside the loop does not help the matter either.

Thinking that maybe it is an issue with the archaic `var` keyword, and using the modern `let` keyword:
```js
for (i = 0; i < GLctx.getProgramParameter(program, 0x8B86/*GL_ACTIVE_UNIFORMS*/); ++i) {
  let u = GLctx.getActiveUniform(program, i),
    nm = u.name,
    sz = u.size,
    lb = getLeftBracePos(nm),
    arrayName = lb > 0 ? nm.slice(0, lb) : nm,
    id = program.uniformIdCounter;

  program.uniformIdCounter += sz;

  // Eagerly get the location of the uniformArray[0] base element.
  // The remaining indices >0 will be left for lazy evaluation to
  // improve performance. Those may never be needed to fetch, if the
  // application fills arrays always in full starting from the first
  // element of the array.
  uniformSizeAndIdsByName[arrayName] = [sz, id];

  // Store placeholder integers in place that highlight that these
  // >0 index locations are array indices pending population.
  for(j = 1; j < sz; ++j) {
    uniformLocsById[++id] = j;
    program.uniformArrayNamesById[id] = arrayName;
  }
}
```
does not help either, and curiously causes Closure to emit a mixture of `var`s and `let`s in the generated code:
```js
for (a.P = d = {}, a.T = {}, h = 0; h < P.getProgramParameter(a, 35718); ++h) {
    var n = P.getActiveUniform(a, h);
    var r = n.name;
    n = n.size;
    var x = c(r);
    x = 0 < x ? r.slice(0, x) : r;
    let u = a.R;
    a.R += n;
    e[x] = [n, u];
    for (r = 1; r < n; ++r) d[++u] = r, a.T[u] = x
}
```

Any thoughts on whether there would be a way to recover the lost 12 bytes on this kind of construct?

I tried to reduce the example down to a standalone test case, but unfortunately was unable to. The test case can be reproduced in Emscripten by running the test `other.test_minimal_runtime_code_size_hello_webgl2_wasm` if anyone is interested in checking it out locally.

**Comment by concavelenz** [comment]
The compiler will collapse "var" statements that do not have statements in between.  "CollapseVariableDeclarations" does this:

https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/CollapseVariableDeclarations.java

https://github.com/google/closure-compiler/blob/master/test/com/google/javascript/jscomp/CollapseVariableDeclarationsTest.java

It is enabled by default here:

https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/CompilationLevel.java#L131


You can see a demonstration of this in action here:

https://closure-compiler-debugger.appspot.com/#input0%3D%250A%250Afunction%2520f()%2520%257B%250A%2520%2520var%2520a%2520%253D%25201%253B%250A%2520%2520var%2520b%2520%253D%25202%253B%250A%250A%2520%2520for%2520(var%2520i%2520%253D%25200%253Bi%253C10%253B)%2520%257B%250A%2520%2520%2520%2520var%2520c%2520%253D%2520'str'%253B%250A%2520%2520%2520%2520var%2520d%2520%253D%2520xxx()%253B%250A%2520%2520%257D%250A%257D%26input1%26conformanceConfig%26externs%3Dvar%2520xxx%253B%26refasterjs-template%26CHECK_TYPES%3Dtrue%26REWRITE_MODULES_BEFORE_TYPECHECKING%3Dtrue%26COLLAPSE_VARIABLE_DECLARATIONS%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue

I don't know how you are running the compiler, do you have a custom runner?

**Comment by juj** [comment]
> I don't know how you are running the compiler, do you have a custom runner?

Yeah, we do. It may be a bit involved to set up, but it can be done so via

```
git clone https://github.com/emscripten-core/emsdk.git
cd emsdk
./emsdk install sdk-upstream-main-64bit
./emsdk activate sdk-upstream-main-64bit
source ./emsdk_env.sh
cd emscripten/main
git remote add juj https://github.com/juj/emscripten.git
git fetch juj
git checkout new_uniform_location_handling
tests/runner other.test_minimal_runtime_code_size_hello_webgl2_wasm

That will execute Closure (obtained via `npm install` as part of `emsdk install` cmd above) and print to stdout the minified results of the command.

**Comment by juj** [comment]
> The compiler will collapse "var" statements that do not have statements in between.

Is `var` initialization a statement in this sense? E.g. `P.getActiveUniform(a, h)` above?

If so, it seems that it would be a bit limiting then, and I'd probably be better off not writing the `var`s in the first place, and manually optimize. Though I wonder what could cause it to introduce `var`s on its own, even when I explicitly left them out?

**Comment by concavelenz** [comment]
No, since the "var" doesn't itself have any side-effects, it can pull together declarations with initializations.  I would look at how you configure the compiler and ensure you have the equivalent to `options.setCollapseVariableDeclarations(true);` or one of the standard option groups (SIMPLE, ADVANCED) setup.

**Comment by juj** [comment]
We run Closure Compiler with

```
C:\emsdk\emscripten\main\node_modules\.bin\google-closure-compiler.cmd --compilation_level ADVANCED_OPTIMIZATIONS --language_in ECMASCRIPT_2020 --language_out NO_TRANSPILE --emit_use_strict=false --externs tmpfxyh_ca6.js --js_output_file tmptkeufgoh.cc.js --js tmpfefs_78u.js
```
That should kick off that optimization?

**Comment by concavelenz** [comment]
I would expect it to, can you attach the externs and js that you are using?

**Comment by juj** [comment]
Yeah, here are the externs and input .js files:

[main.zip](https://github.com/google/closure-compiler/files/6234903/main.zip)

Executed in the test suite with

```
C:\emsdk\emscripten\main\node_modules\.bin\google-closure-compiler.cmd --compilation_level ADVANCED_OPTIMIZATIONS --language_in ECMASCRIPT_2020 --language_out NO_TRANSPILE --emit_use_strict=false --externs externs.js --js_output_file output.js --js input.js
```

**Comment by brad4d** [comment]
What's happening here is an interaction between a pass that tries to minimize the total number of variables by re-using variable names and the pass that tries to gather all of the variables together and declare them in one statement.

Original code:
```js
for (i = 0; i < GLctx.getProgramParameter(program, 0x8B86/*GL_ACTIVE_UNIFORMS*/); ++i) {
  var u = GLctx.getActiveUniform(program, i);
  var nm = u.name;
  var sz = u.size;
  var lb = getLeftBracePos(nm);
  var arrayName = lb > 0 ? nm.slice(0, lb) : nm;

  // Assign a new location.
  var id = program.uniformIdCounter;
  program.uniformIdCounter += sz;

  // Eagerly get the location of the uniformArray[0] base element.
  // The remaining indices >0 will be left for lazy evaluation to
  // improve performance. Those may never be needed to fetch, if the
  // application fills arrays always in full starting from the first
  // element of the array.
  uniformSizeAndIdsByName[arrayName] = [sz, id];

  // Store placeholder integers in place that highlight that these
  // >0 index locations are array indices pending population.
  for(j = 1; j < sz; ++j) {
    uniformLocsById[++id] = j;
    program.uniformArrayNamesById[id] = arrayName;
  }
}
```

Notice that the `u` variable isn't used again after the 3rd line of the loop.
The compiler decides to re-use that variable name instead of having a separate `sz` variable.
Thus in the output (below) we see the variable `n` (renaming of `u`) being assigned twice.
It makes the same decision for `lb` and `arrayName`.
Normally such decisions are a size win, but here they prevent the variable declarations from being combined.

```js
for (a.P = d = {}, a.T = {}, h = 0; h < P.getProgramParameter(a, 35718); ++h) {
    var n = P.getActiveUniform(a, h);
    var r = n.name;
    n = n.size;
    var w = c(r);
    w = 0 < w ? r.slice(0, w) : r;
    var F = a.R;
    a.R += n;
    e[w] = [n, F];
    for (r = 1; r < n; ++r) d[++F] = r, a.T[F] = w
}
```

The `n` and `r` variables are declared in separate statements because the definition of `r` uses `n`.
The definition of `w` gets its own statement because it calls a function that might (theoretically) be affected by the assignment to `n`.
The declaration of `F` gets its own statement because it uses the value of `a.R`, which could have been affected by the `c(r)` function call or possibly the `slice()` call. (Yes, as humans we know `slice()` won't affect anything, but the compiler isn't necessarily that smart.)

So to sum up, this is caused by an unfortunate interaction of individual optimizations.
It doesn't seem worthwhile to dig into this further. The output code works, and the interaction is a fairly rare one, so not a general problem for code size.
